pkg/httpstream/spdy/upgrade.go

```diff
@@ -0,0 +1,127 @@
+package spdy
+
+import (
+	"bufio"
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/httpstream"
+	"github.com/docker/spdystream"
+)
+
+const (
+	headerConnection = "Connection"
+	headerUpgrade    = "Upgrade"
+	headerSpdy31     = "SPDY/3.1"
+)
+
+// connectionUpgrader knows how to upgrade HTTP requests and responses. It
+// implements the httpstream.ConnectionUpgrader interface.
+type connectionUpgrader struct {
+}
+
+// NewConnectionUpgrader returns a new httpstream.ConnectionUpgrader that is
+// capable of upgrading HTTP requests and responses using SPDY/3.1 via the
+// spdystream package.
+func NewConnectionUpgrader() httpstream.ConnectionUpgrader {
+	return connectionUpgrader{}
+}
+
+// UpgradeRequest upgrades an HTTP request to one that supports multiplexed
+// streams. newStreamHandler will be called synchronously whenever the
+// other end of the upgraded connection creates a new stream.
+//
+// UpgradeRequest uses req to create its own net.Connection to the remote
+// server.
+func (u connectionUpgrader) UpgradeRequest(req *http.Request, newStreamHandler httpstream.NewStreamHandler) (httpstream.Connection, error) {
+	req.Header.Add(headerConnection, headerUpgrade)
+	req.Header.Add(headerUpgrade, headerSpdy31)
+
+	// TODO ideally we'd tie in to the Kubernetes Client.Do() but there's no easy
```

Comment:
I don't think hijack is ever going to work, so we need to make sure client.Config can represent all the info you need to do what you need.  Fortunately, it does.  There should be a NewUpgrader(client.Config) that returns an object you can do a standard Get on (with all the defaulting that implies), and hands you back the connection.